### PR TITLE
ECP-756: GraphQL Schema for Rewards

### PR DESCRIPTION
**Pull request by paliarush**
https://jira.corp.magento.com/browse/ECP-756

### CODE REVIEW

File: design-documents/graph-ql/coverage/reward-points.graphqls [hunk 1]
```diff
@@ -0,0 +1,38 @@
+type Customer {
+    reward_points: RewardPoints @doc(description: "Customer reward points")
+}
+
+type RewardPoints {
+    balance: RewardPointsBalance @doc(description: "Current reward points balance")
+    exchange_rate: RewardPointsExchangeRate @doc(description: "Current reward points exchange rate")
+    subscription_status: RewardPointsSubscriptionStatus @doc(description: "Reward points related email subscription status")
+    balance_history: [RewardPointsBalanceHistoryItem] @doc(description: "Reward points balance history. Visibility of this section depends on the configuration")
```
Comment (DrewML) [inline design-documents/graph-ql/coverage/reward-points.graphqls:9]: > Visibility of this section depends on the configuration

This comment doesn't make it clear whether the type definition will not exist in the schema based on config, or if it'll just be `null` depending on config

Comment (paliarush) [reply design-documents/graph-ql/coverage/reward-points.graphqls]: Thank you, please check adjusted description.

Comment (DrewML) [reply design-documents/graph-ql/coverage/reward-points.graphqls]: Thanks! New description is much more clear

File: design-documents/graph-ql/coverage/reward-points.graphqls [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+    reward_points: RewardPoints @doc(description: "Customer reward points details")
+}
+
+type Cart {
+    applied_reward_points: RewardPointsAmount @doc(description: "Reward points ammount applied to the cart")
+}
+
+type Mutation {
+    applyRewardPointsToCart(cartId: ID!): ApplyRewardPointsToCartOutput @doc(description: "Apply all available points up to the cart total. Partial redemption is not available")
+    removeRewardPointsFromCart(cartId: ID!): RemoveRewardPointsFromCartOutput @doc(description: "Cancel usage of the previously applied reward points to cart")
```
Comment (DrewML) [inline design-documents/graph-ql/coverage/reward-points.graphqls:11]: Probably not the end of the world, but the symmetry here seems off with the naming

- `apply`
- `remove`

I think `apply` makes sense, since you're not conceptually _adding_ to the cart.

But if you're not `adding` something, you wouldn't expect to be able to `remove` that thing.

tl;dr: I think `apply` and `remove` don't work well together, but I'm not sure of a better alternative

File: design-documents/graph-ql/coverage/reward-points.graphqls [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+    money: Money! @doc(description: "Reward points amount in store currency")
+}
+
+type RewardPointsExchangeRates @doc (description: "Exchange rates depend on the customer group"){
+    earning: Float! @doc(description: "How many points are earned per 1 currency spent")
+    redemption: Float! @doc(description: "How many points need to be redeemed to get 1 currency discount at the checkout")
+}
+
+type RewardPointsSubscriptionStatus {
+    balance_updates: SubscriptionStatus! @doc(description: "Customer subscription status to 'Reward points balance updates' emails")
```
Comment (DrewML) [inline design-documents/graph-ql/coverage/reward-points.graphqls:40]: `balance_updates` seems a bit generic here, while the return type and the description give a lot more context.

Can we name this field in a way that it's less likely folks need to reference the docs? Maybe `balance_update_email_subscription`? Going to be a trade-off with verbosity, maybe we can find a happy middle ground.

File: design-documents/graph-ql/coverage/reward-points.graphqls (hunk 2 above)
Comment (paliarush) [reply design-documents/graph-ql/coverage/reward-points.graphqls:11]: This the the naming we currently have for coupons and gift cards.. not sure whether there is a good alternative. Also seems better to keep similar naming for consistency.

File: design-documents/graph-ql/coverage/reward-points.graphqls (hunk 3 above)
Comment (paliarush) [reply design-documents/graph-ql/coverage/reward-points.graphqls:40]: The type name already contains "SubscriptionStatus", so if you go from there it is "balance updates subscription status". Let me know if you still believe it is better to change it.

File: design-documents/graph-ql/coverage/reward-points.graphqls (hunk 2 above)
Comment (DrewML) [reply design-documents/graph-ql/coverage/reward-points.graphqls:11]: > Also seems better to keep similar naming for consistency.

I think this could be argued in both directions - depends on whether you're optimizing for folks new to the API, or folks familiar with the existing API surface.

But I also can't think of better wording so it's a moot point 😂. Makes sense to me to just move forward by matching what we have 👍

File: design-documents/graph-ql/coverage/reward-points.graphqls [hunk 4]
```diff
@@ -0,0 +1,70 @@
...
+    applied_reward_points: RewardPointsAmount @doc(description: "Reward points ammount applied to the cart")
+}
+
+type Mutation {
+    applyRewardPointsToCart(cartId: ID!): ApplyRewardPointsToCartOutput @doc(description: "Apply all available points up to the cart total. Partial redemption is not available")
+    removeRewardPointsFromCart(cartId: ID!): RemoveRewardPointsFromCartOutput @doc(description: "Cancel usage of the previously applied reward points to cart")
+}
+
+type ApplyRewardPointsToCartOutput {
+    cart: Cart
```
Comment (DrewML) [inline design-documents/graph-ql/coverage/reward-points.graphqls:15]: We've discussed that, for object types, we want non-scalar fields to be nullable, which we've followed here.

But, for mutation return types, I'm wondering if we expect a case where we could do a _write_ to the cart service, but then not be able to do a read?

The reason for nullable on non-scalar types is to guard ourselves from cascading failures when some services are down. Do we have concerns that cross service boundaries here (honest question, I truly do not know)?

One trade-off we make with nullability is that the client-side has to do conditional checks when it gets a response, and handle additional states.

```js
function MyComponent() {
   // ... run mutation here
   if (response.error) {
      // handle error updating cart
   }
   if (response.cart) {
      // safe to use cart data when rendering
   } else {
      // handle case where we were able to successfully update the cart,
      // but weren't able to query for new results
   }
}
```

when we go non-nullable, that can turn into:

```js
function MyComponent() {
   if (response.error) {
      // handle error updating cart
   }
   
   // render UI with cart data
}
```

I think non-nullable makes a bit more sense in the context of a mutation return type since we have more context about what the client would expect. Thoughts?

File: design-documents/graph-ql/coverage/reward-points.graphqls (hunk 3 above)
Comment (DrewML) [reply design-documents/graph-ql/coverage/reward-points.graphqls:40]: That's a good point - I think it's good as-is 👍

File: design-documents/graph-ql/coverage/reward-points.graphqls [hunk 5]
```diff
@@ -0,0 +1,70 @@
...
+
+type RewardPointsBalanceHistoryItem {
+    balance: RewardPointsAmount @doc(description: "Reward points balance after the completion of the transaction")
+    points_change: Float! @doc(description: "Number of points added or deducted in the transaction")
+    change_reason: String! @doc(description: "Reason for balance change")
+    date: String! @doc(description: "Transaction date")
+}
+
+type StoreConfig {
+    magento_reward_general_is_enabled: String @doc(description: "Reward points functionality status: enabled/disabled")
```
Comment (DrewML) [inline design-documents/graph-ql/coverage/reward-points.graphqls:57]: Is there something in core preventing us from using a Boolean here?

File: design-documents/graph-ql/coverage/reward-points.graphqls (hunk 4 above)
Comment (paliarush) [reply design-documents/graph-ql/coverage/reward-points.graphqls:15]: Sounds reasonable, fixing it here. @DrewML , can I ask you to write these rules down somewhere in https://github.com/magento/architecture/tree/master/design-documents/graph-ql ?

1. Non-nullable for mutations
1. Nullable for objects in queries
1. Non-nullable for scalars in queries

File: design-documents/graph-ql/coverage/reward-points.graphqls (hunk 5 above)
Comment (paliarush) [reply design-documents/graph-ql/coverage/reward-points.graphqls:57]: The logic here is, we take data from `core_config_data` as is and run it through the list of fields declared in this GraphQL schema to leave only allowed ones. 
These values are currently stored in DB as 0/1. I don't think GraphQL framework will properly convert them to boolean. It may be possible, but will require additional effort to add support.

Comment (DrewML) [reply design-documents/graph-ql/coverage/reward-points.graphqls:57]: I can definitely see how this makes the implementation easier/lower effort.

My concern is that the work doesn't go away, we're just pushing it to every client instead. Every client now has to do:

```js
if (magento_reward_general_is_enabled === 'enabled') {

}
```

rather than us returning a more descriptive type (`Boolean`)

Comment (fnhipster) [reply design-documents/graph-ql/coverage/reward-points.graphqls:57]: My 2 cents: there is already a lot of unnecessary front-end logic in the client. e.g. mapping and transforming data coming from GraphQL. Although this might seem small, it does make a more descriptive and elegant API. We gotta start trimming logic where we can in order to enable clients to be leaner.

Comment (paliarush) [reply design-documents/graph-ql/coverage/reward-points.graphqls:57]: I would address it in scope of separate story related to `StoreConfig`. At the time it was implemented we agreed to the tradeoff. 
Otherwise Reward implementation will be blocked by the rework of `StoreConfig`. 

@nrkapoor are you okay with creating a separate story for improving DX of `StoreConfig` query?

Comment (nrkapoor) [reply design-documents/graph-ql/coverage/reward-points.graphqls:57]: @paliarush Yes I will create a separate ticket for storeconfig clean.

Comment (DrewML) [reply design-documents/graph-ql/coverage/reward-points.graphqls:57]: Sounds good - thanks y'all 👍

File: design-documents/graph-ql/coverage/reward-points.graphqls (hunk 4 above)
Comment (DrewML) [reply design-documents/graph-ql/coverage/reward-points.graphqls]: Thanks for updating!

Left myself a TODO to add a document covering our experiences/preferences with nullability

**Comment by DrewML** [comment]
Great work @paliarush!
